cli_tools/gce_vm_image_import/importer/inflater.go

```diff
@@ -40,6 +43,119 @@ type inflater interface {
 	traceLogs() []string
 }
 
+// inflaterFacade implements an inflater using other concrete implementations.
+type inflaterFacade struct {
+	mainInflater   inflater
+	shadowInflater inflater
+}
+
+// signals to control the verification towards shadow inflater
+const (
+	sigMainInflaterDone   = "main done"
+	sigMainInflaterErr    = "main err"
+	sigShadowInflaterDone = "shadow done"
+	sigShadowInflaterErr  = "shadow err"
+)
+
+func (facade inflaterFacade) inflate(ctx context.Context) (persistentDisk, error) {
+	inflaterChan := make(chan string)
+
+	// Launch main inflater.
+	var pd persistentDisk
+	var err error
+	go func() {
+		pd, err = facade.mainInflater.inflate(ctx)
+		if err != nil {
+			inflaterChan <- sigMainInflaterErr
+		} else {
+			inflaterChan <- sigMainInflaterDone
+		}
+	}()
+
+	if facade.shadowInflater == nil {
+		pd.matchResult = "NA"
+		// Wait for main inflater to finish
+		<-inflaterChan
+		return pd, err
+	}
+
+	// Launch shadow inflater.
+	var shadowPd persistentDisk
+	var shadowErr error
+	go func() {
+		shadowPd, shadowErr = facade.shadowInflater.inflate(ctx)
+		if shadowErr != nil {
+			inflaterChan <- sigShadowInflaterErr
+		} else {
+			inflaterChan <- sigShadowInflaterDone
+		}
+	}()
+
+	// Return early if main inflater finished first.
+	result := <-inflaterChan
+	if result == sigMainInflaterDone || result == sigMainInflaterErr {
+		if result == sigMainInflaterDone {
+			pd.matchResult = "Main inflater finished earlier"
+		}
+		return pd, err
+	}
+
+	// Wait for main inflater to finish, then process shadow inflater's result.
+	mainResult := <-inflaterChan
+	if result == sigShadowInflaterDone {
+		if mainResult == sigMainInflaterErr {
+			pd.matchResult = "Main inflater failed while shadow inflater succeeded"
+		} else {
+			facade.compareWithShadowInflater(ctx, &pd, &shadowPd)
+		}
+	} else if result == sigShadowInflaterErr && mainResult == sigMainInflaterDone {
+		if isCausedByUnsupportedFormat(shadowErr) {
+			pd.matchResult = "Shadow inflater doesn't support the format while main inflater supports"
+		} else {
+			pd.matchResult = "Shadow inflater failed while main inflater succeeded"
+		}
+	}
```

Comment:
Are we keeping enough logs to help the PD team debug the native VMDK failures?

Reply:
PD has their own log, everything we got here is sent from their side. We just need to let then know the rough time range.